### DISCUSSION TITLE
How to allow for check if API key exists in the api_keys table, but not allow full public read access to entire table?

### DISCUSSION THREAD

**Discussion by pgibler** [supabase/supabase#21045] [Questions] [answered]
I have a table that stores API keys. It has an api_key and a user_id. Setting up the update / insert policy's using RLS was simple enough. However, for read access, I only want the api_key value to be readable. And I want to prevent a `select api_key from api_keys` query from being runnable.

Essentially, the only query I want to allow is
`select api_key from api_keys where api_key=<input api key>`
Or some equivalent to achieve what I'm looking for in the title.

How would I do this?

For reference, here's the Typescript code I'm using to generate & store + retrieve the key:

```
export const generateAndStoreApiKey = async (user: User): Promise<string> => {
  const newApiKey = uuidv4();
  const { error } = await supabase
    .from('api_keys')
    .upsert({ user_id: user.id, api_key: newApiKey });

  if (error) throw error;
  return newApiKey;
};

export const retrieveApiKey = async (user: User) => {
  const { data, error } = await supabase
    .from('api_keys')
    .select('api_key')
    .eq('user_id', user.id)
    .maybeSingle();

  if (error) throw error;
  return data;
};
```

**Comment by mwoss** [comment, marked as answer]
I would suggest looking at Column Level Security - https://supabase.com/docs/guides/auth/column-level-security. By using CLS you could disallow lookup for `user_id` for all users. 

Btw I hope that you encrypt API keys before storing them in DB. Please do not store bare values there, it's a big security concern.

**Reply by mwoss** [reply]
One more option - disallow access to this table for all users (RLS off) and use Postgres functions to create your API where you can control what data is returned.

**Reply by pgibler** [reply]
Thank you for the suggestion on using CLS mwoss. I'll also look into creating a postgres function to handle it.

Good call on the encryption of the keys - I hadn't thought of that yet and you're right, I'll have to do that before launching.
